### PR TITLE
Add Fast Image Processor for mobileViT

### PR DESCRIPTION
**Pull request by MinJu-Ha**
related to #36978
cc: @yonigozlan

I added Fast image processor for mobileViT and I noticed a noticeable difference between the outputs after preprocessing.

Here’s the code I used to compare them:

```
diff = (encoding_slow.pixel_values - encoding_fast.pixel_values).abs()
print(f"\n📊 Difference statistics:")
print(f"  Max difference: {diff.max().item():.10f}")
print(f"  Mean difference: {diff.mean().item():.10f}")
print(f"  Slow min/max: {encoding_slow.pixel_values.min().item():.10f} ~ {encoding_slow.pixel_values.max().item():.10f}")
print(f"  Fast min/max: {encoding_fast.pixel_values.min().item():.10f} ~ {encoding_fast.pixel_values.max().item():.10f}")
print(f"Slow implementation dtype: {encoding_slow.pixel_values.dtype}")
print(f"Fast implementation dtype: {encoding_fast.pixel_values.dtype}")
```
results:
📊 Difference statistics: 
  Max difference: 0.3411765397
  Mean difference: 0.1117687449
  Slow min/max: 0.0313725509 ~ 0.9764705896
  Fast min/max: 0.0313725509 ~ 0.9764706492
Slow implementation dtype: torch.float32
Fast implementation dtype: torch.float32
```

Even though the size configs look the same ({'shortest_edge': 20}), and both use torch.float32, the output difference seems quite significant for a slow/fast equivalence test.

### CODE REVIEW

**Review by yonigozlan — COMMENTED** [review]
Hi @MinJu-Ha ! Thanks for contributing this! Look slike you're missing the `do_flip_channel_order` arg that is present in the slow image processor, which could explain the differences you're getting (We should have ~1e-05 in mean diff). Could you try to implement the necessarry changes? Thanks!

**Comment by MinJu-Ha** [comment]
It really was—thank you so much for the advice! I'm still working on this issue and have finally passed 18 tests. This week, I’ll be focusing on the remaining 2 skipped ones, which are related to batched image processing and CUDA.

**Comment by MinJu-Ha** [comment]
@yonigozlan, waiting for your review!

**Comment by MinJu-Ha** [comment]
@yonigozlan , I'm still waiting for your review!

and just to clarify:
The test failures in `tests/models/mistral3/test_processor_mistral3.py` seem to be caused by external image download issues (e.g., PIL.UnidentifiedImageError when loading from https://www.ilankelman.org/stopsigns/australia.jpg).
Since my changes don’t touch this file or its logic, I believe these failures are unrelated to my PR.

Let me know if you’d like me to skip the tests or mock the image loading as a workaround!

image

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+    def flip_channel_order(self, image):
+        # Check if we have 3 or more channels
+        if image.shape[0] >= 3:
+            # Flip only the first 3 channels (RGB → BGR)
+            flipped = image.clone()
+            flipped[0:3] = image[[2, 1, 0], ...]
+            return flipped
+        # For grayscale or other formats, return as is
+        return image
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py:62]: This is not used

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    # If the image processor requires more complex augmentations, methods from BaseImageProcessorFast can be overridden.
+    # In most cases, only the `_preprocess` method should be overridden.
+
+    # For an example of a fast image processor requiring more complex augmentations, see `LlavaNextImageProcessorFast`.
+
+    # Default values should be checked against the slow image processor
+    # None values left after checking can be removed
+    resample = PILImageResampling.BILINEAR
+    image_mean = None
+    image_std = None
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py:43]: You can remove the None values
```suggestion
```

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        processed_images = reorder_images(processed_images_grouped, grouped_images_index)
+
+        # Stack all processed images if return_tensors is specified
+        if return_tensors is not None:
+            processed_images = torch.stack(processed_images, dim=0)
+
+        return BatchFeature(data={"pixel_values": processed_images}, tensor_type=return_tensors)
+
+    def post_process_semantic_segmentation(self, *args, **kwargs):
+        raise NotImplementedError("This method is not implemented for MobileViTImageProcessorFast.")
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py:136]: You need to implement it

File: tests/models/mobilevit/test_image_processing_mobilevit.py [hunk 4]
```diff
@@ -119,12 +123,13 @@ def image_processor_dict(self):
...
+        for image_processing_class in self.image_processor_list:
+            image_processing = image_processing_class(**self.image_processor_dict)
+            self.assertTrue(hasattr(image_processing, "do_resize"))
+            self.assertTrue(hasattr(image_processing, "size"))
+            self.assertTrue(hasattr(image_processing, "do_center_crop"))
+            self.assertTrue(hasattr(image_processing, "center_crop"))
+            self.assertTrue(hasattr(image_processing, "do_flip_channel_order"))
 
     def test_image_processor_from_dict_with_kwargs(self):
         image_processor = self.image_processing_class.from_dict(self.image_processor_dict)
```
Comment (yonigozlan) [inline tests/models/mobilevit/test_image_processing_mobilevit.py:135]: This test and the following one need to include both image processors

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            flipped = image.clone()
+            flipped[0:3] = image[[2, 1, 0], ...]
+            return flipped
+        # For grayscale or other formats, return as is
+        return image
+
+    def _preprocess(
+        self,
+        images,
+        do_resize=True,
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py:67]: You should handle segmentation_maps here as well, you can do it by overriding `preprocess` if that's easier. Have a look at how it's done for Beit fast image processor.

**Review by yonigozlan — COMMENTED** [review]
Hey @MinJu-Ha , thanks for iterating :) . You are still missing the logic for handling segmentation_maps

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py (hunk 5 above)
Comment (MinJu-Ha) [reply src/transformers/models/mobilevit/image_processing_mobilevit_fast.py]: Thanks! I overrode the `preprocess` method based on BEiT.
I’d appreciate it if you could review my changes when you have some time.

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+                # For batched images, we need to handle them all at once
+                if stacked_images.ndim > 3 and stacked_images.shape[1] >= 3:
+                    # Flip RGB → BGR for batched images
+                    flipped = stacked_images.clone()
+                    flipped[:, 0:3] = stacked_images[:, [2, 1, 0], ...]
+                    stacked_images = flipped
+            if do_normalize and image_mean is not None and image_std is not None:
+                stacked_images = self.normalize(image=stacked_images, mean=image_mean, std=image_std)
+            if do_convert_rgb:
+                stacked_images = self.convert_to_rgb(stacked_images)
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py]: No need for that as they are not present in the slow image processor
```suggestion
```

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+            processed_images_grouped[shape] = stacked_images
+
+        processed_images = reorder_images(processed_images_grouped, grouped_images_index)
+
+        # Stack all processed images if return_tensors is specified
+        if return_tensors is not None:
+            processed_images = torch.stack(processed_images, dim=0)
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py]:
```suggestion
        processed_images = torch.stack(processed_images, dim=0) if return_tensors else processed_images
```

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+            )
+
+            if segmentation_map.ndim == 2:
+                segmentation_map = segmentation_map[None, ...]
+
+            processed_segmentation_maps.append(segmentation_map)
+
+        kwargs["do_normalize"] = False
+        kwargs["do_rescale"] = False
+        kwargs["input_data_format"] = ChannelDimension.FIRST
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py]: No need for thesee (typo in beit)
```suggestion
        kwargs["do_rescale"] = False
```

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        do_rescale=True,
+        rescale_factor=None,
+        do_center_crop=True,
+        crop_size=None,
+        do_flip_channel_order=True,
+        do_convert_rgb=False,
+        return_tensors=None,
+        do_normalize=None,
+        image_mean=None,
+        image_std=None,
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py]: let's not have default value for args in _preprocess, like in other fast image processors. Also would be nice to type these args. you don't need to have do_normalize, image_mean, image_std here. Just add a **kwargs at the end of the signature

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    ):
+        processed_images = []
+
+        if do_normalize is None:
+            do_normalize = self.do_normalize
+        if image_mean is None and hasattr(self, "image_mean"):
+            image_mean = self.image_mean
+        if image_std is None and hasattr(self, "image_std"):
+            image_std = self.image_std
```
Comment (yonigozlan) [inline src/transformers/models/mobilevit/image_processing_mobilevit_fast.py]:
```suggestion
```

File: tests/models/mobilevit/test_image_processing_mobilevit.py [hunk 11]
```diff
@@ -18,7 +18,7 @@
 from datasets import load_dataset
```
Comment (yonigozlan) [inline tests/models/mobilevit/test_image_processing_mobilevit.py:19]: Let's also have slow_fast equivalence tests for segmentation maps by overriding `test_slow_fast_equivalence` and `test_slow_fast_equivalence_batched` in `MobileViTImageProcessingTest`

**Review by yonigozlan — COMMENTED** [review]
Hi @MinJu-Ha , thanks for iterating, it's looking a lot better! Still left some things to modify and some tests to add, but almost ready yo merge :)

**Comment by yonigozlan** [comment]
Hello @MinJu-Ha , don't hesitate to ping me if you need help or if the PR is ready for review!

File: src/transformers/models/mobilevit/image_processing_mobilevit_fast.py (hunk 9 above)
Comment (MinJu-Ha) [reply src/transformers/models/mobilevit/image_processing_mobilevit_fast.py]: Thank you for your advice! I reflected in commit `add **kwargs and remove default values in _preprocess`

**Comment by MinJu-Ha** [comment]
Hi @yonigozlan ! Sorry for the delay — I've been busy recently due to my U.S. visa process.  
Thanks so much for your patience!  
I've just finished reflecting your suggestions in the code.  
Please have a look when you have time, and let me know if any further updates are needed!

**Review by yonigozlan — APPROVED** [review]
Thanks @MinJu-Ha for iterating on this! Had to make some small changes mainly because of recent updates in Transformers, but LGTM! Waiting for the PR to be green then I'll merge
